### PR TITLE
Fix array equality check , clean 'logger.finer' statements

### PR DESCRIPTION
**Pull request by DanVanAtta**
Static analysis flagged two arrays being compared against each other (equality match as opposed to logical). Turns out we do this as a way to check if both arrays are null, in both cases we have a comment saying so. This is a bit 'overly clever', we can instead dumb it down and check if both variables are not null, and remove the comment.

Along the way there were a few logger.fine statements that could get removed.

### CODE REVIEW

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java [hunk 1]
```diff
@@ -14,16 +14,13 @@
...
     final Method[] methods = remoteInterface.getMethods();
     Arrays.sort(methods, methodComparator);
-    if (logger.isLoggable(Level.FINEST)) {
-      logger.fine("Sorted methods:" + Arrays.asList(methods));
-    }
+
     for (int i = 0; i < methods.length; i++) {
       if (methods[i].getName().equals(methodName)) {
         final Class<?>[] types = methods[i].getParameterTypes();
-        // both null
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Is all this code really just trying to express the following?

```java
if (Arrays.equals(types, argTypes)) {
  return i;
}
```

Tests in `RemoteInterfaceHelperTest` still pass, although there's no coverage on the throw `IllegalStateException` case.

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java [hunk 2]
```diff
@@ -61,8 +55,7 @@ static int getNumber(final String methodName, final Class<?>[] argTypes, final C
     }
     final Class<?>[] t1 = o1.getParameterTypes();
     final Class<?>[] t2 = o2.getParameterTypes();
-    // both null
-    if (t1 == t2) {
+    if ((t1 == null) && (t2 == null)) {
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: This complete method/lambda can be replaced with
```java
Comparator.comparing(Method::getName)
  .thenComparing(Method::getParameterTypes,
    Comparator.comparing(o -> o == null)
    .thenComparingInt(a -> a.length)
    .thenComparing((o1, o2) -> {/*insert for-loop array comparison*/});

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 1 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Seems like it.

Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Looks so : )

- 19f69e80643adda8dbe38d2fd6ccb9088753439f rewrites the array equality check
- 737a7cbde3f82fbe381830031bf5612eafde253e then rewrites with stream notation

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java [hunk 3]
```diff
@@ -3,49 +3,25 @@
...
-        }
-      }
-    }
-    throw new IllegalStateException("Method not found");
+
+    return IntStream.range(0, methods.length)
+        .filter(i -> methods[i].getName().equals(methodName))
+        .filter(i -> Arrays.equals(argTypes, methods[i].getParameterTypes()))
+        .findFirst()
+        .orElseThrow(() -> new IllegalStateException("Method not found: " + methodName));
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: 👍

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 2 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: It seems the last thenComparing bit is not required.

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Or that the last thenComparing bit needs to be
`.thenComparing(Function.identity())`

Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Trying this out:
```
  private static final Comparator<Method> methodComparator = Comparator.comparing(Method::getName)
      .thenComparing(Method::getParameterTypes, typeName ->
          Comparator.comparing(o -> o == null)
              .thenComparingInt(a -> a.length)
              .thenComparing((o1, o2) -> {/*insert for-loop array comparison*/});

```
I get a compilation error on `a.length`

Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Fixed the bad arrow: length is wrong type.
```
  private static final Comparator<Method> methodComparator = Comparator.comparing(Method::getName)
      .thenComparing(Method::getParameterTypes,
          Comparator.comparing(o -> o == null)
              .thenComparingInt(a -> a.length)
              .thenComparing(Function.identity())
              .thenComparing((o1, o2) -> {/*insert for-loop array comparison*/});
```

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Perhaps using simply `thenComparing` instead of `thenComparingInt`?

Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Hmm,  gave a few things a try, no luck. Can pick back up later. One caution, we need to be careful about current null handling, I wonder if comparator could change that on us?

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 3 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: @DanVanAtta looking at this again it seems like this method is always returning `methods.length - 1` or throwing an exception, so the intstream could be mapped to a method object to simply the code

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 2 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: The thenComparing statements are short-circuiting AFAIK.
Couldn't find this explicitly mentioned anywhere, but at least the OpenJDK implementation is.

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: I wonder if
```java
Comparator.comparing(Method::getName)
  .thenComparing(Method::getParameterTypes);
``` 
would be sufficient for the sorting order.

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: It also seems `Method#getParameterTypes` doesn't return 0

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: It seems we need to explicitly specify the type argument:
```java
Comparator.comparing(Method::getName)
  .thenComparing(Method::getParameterTypes,
    Comparator.comparingInt((Class<?>[] a) -> a.length)
    .thenComparing((o1, o2) -> {
      for (int i = 0; i < o1.length; i++) {
        final int compareValue = o1[i].getName().compareTo(o2[i].getName());
        if (compareValue != 0) {
          return compareValue;
        }
        return 0;
      }));
```
This snippet works for me

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 3 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Nevermind, it doesn't

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java [hunk 4]
```diff
@@ -3,49 +3,25 @@
...
-          }
-        }
-      }
-    }
-    throw new IllegalStateException("Method not found");
+
+    return IntStream.range(0, methods.length)
+        .filter(i -> methods[i].getName().equals(methodName))
+        .filter(i -> Arrays.equals(argTypes, methods[i].getParameterTypes()))
+        .findFirst()
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: You might want to switch to findAny instead of findFirst in case we want to use parallel streams at some point, since the order should really not matter as there shouldn't be more than one element in this stream anyways

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 2 above)
Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Hmm, I'm not having much luck! :disappointed: 

```

  private void sortMethods(Method[] methods) {
    Arrays.sort(methods, Comparator.comparing(Method::getName)
        .thenComparing(Method::getParameterTypes,
            Comparator.comparingInt((Class<?>[] a) -> a.length)
                .thenComparing((o1, o2) -> {
                  for (int i = 0; i < o1.length; i++) {
                    final int compareValue = o1[i].getName().compareTo(o2[i].getName());
                    if (compareValue != 0) {
                      return compareValue;
                    }
                    return 0;
                  }
                  })));
```
screenshot from 2018-01-02 21-52-45

I'm pretty tempted to just leave this method alone to move on for now, it's okay as is...  : |

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 4 above)
Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: no objection  here, updated:   d668db9

**Comment by DanVanAtta** [comment]
@RoiEXLab @simon33-2  ; let's cut scope on this specific update for now and do a final review pass and merge. The other method can be done on its own branch next or in parallel. If it's really nagging someone, I will accept a PR into this branch : )    I am starting to think  we may be best off to invest effort elsewhere for now - lot's of other things to do and many other low hanging fruit.

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java (hunk 2 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: No problem, I'll submit a follow-up PR.

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Unless, of course, @ssoloff figures out a better way that doesn't require that much explicit typisation like the last time when I insited that this type specification was absolutely required 😬 ^^
